22.0.0.12 post (#2759)
OpenLiberty/blogs: posts/2022-11-22-22.0.0.12.adoc

---
layout: post
title: "CVE and many other notable bug fixes in Open Liberty 22.0.0.12"
# Do NOT change the categories section
categories: blog
author_picture: https://avatars3.githubusercontent.com/mbroz2
author_github: https://github.com/mbroz2
seo-title: CVE and many other notable bug fixes in Open Liberty 22.0.0.12 - OpenLiberty.io
seo-description: Open Liberty 22.0.0.12 provides many significant bug fixes, including one that addresses a CVE in GraphQL Java.  Cloud-hosted version of the "Running GraphQL queries and mutations using a GraphQL client" guide has also been made available.
blog_description: Open Liberty 22.0.0.12 provides many significant bug fixes, including one that addresses a CVE in GraphQL Java.  Cloud-hosted version of the "Running GraphQL queries and mutations using a GraphQL client" guide has also been made available.
open-graph-image: https://openliberty.io/img/twitter_card.jpg
open-graph-image-alt: Open Liberty Logo
---
= CVE and many other notable bug fixes in Open Liberty 22.0.0.12
Michal Broz <https://github.com/mbroz2>
:imagesdir: /
:url-prefix:
:url-about: /
//Blank line here is necessary before starting the body of the post.

> Open Liberty 22.0.0.12 provides many significant bug fixes, including one that addresses a CVE in GraphQL Java.  If you're interested in features and functionality which we have in progress, like Jakarta EE 10 and MicroProfile 6, take a look at the recent link:https://openliberty.io/blog/?search=beta&key=tag[beta blog posts].

Author: ```suggestion
Open Liberty 22.0.0.12 provides many significant bug fixes, including one that addresses a CVE in GraphQL Java.  If you're interested in features and functionality which we have in progress, like Jakarta EE 10, MicroProfile 6 and InstantOn, take a look at the recent link:https://openliberty.io/blog/?search=beta&key=tag[beta blog posts].
```

[#CVEs]
== Security vulnerability (CVE) fixes in this release
[cols="5*"]
|===
|CVE |CVSS Score |Vulnerability Assessment |Versions Affected |Notes

|http://cve.mitre.org/cgi-bin/cvename.cgi?name=CVE-2022-37734[CVE-2022-37734]
|7.5
|Denial of service
|17.0.0.3 - 22.0.0.11
|Affects the link:{url-prefix}/docs/latest/reference/feature/mpGraphQL-1.0.html[mpGraphQL-1.0] and link:{url-prefix}/docs/latest/reference/feature/mpGraphQL-2.0.html[mpGraphQL-2.0] features
|===

For a list of past security vulnerability fixes, reference the link:{url-prefix}/docs/latest/security-vulnerabilities.html[Security vulnerability (CVE) list].

[#bugs]
== Notable bugs fixed in this release

We’ve spent some time fixing bugs. The following sections describe just some of the issues resolved in this release. If you’re interested, here’s the link:https://github.com/OpenLiberty/open-liberty/issues?q=label%3Arelease%3A220012+label%3A%22release+bug%22[full list of bugs fixed in 22.0.0.12].

* link:https://github.com/OpenLiberty/open-liberty/issues/23017[MP Reactive Messaging: NullPointerException during Kafka partition rebalance]
+

> During partition reassignment, it's possible for the Kafka connector for MicroProfile Reactive Messaging to not update its internal state correctly, resulting in it stopping receiving messages and a `NullPointerException` being logged.

Author: ```suggestion
During partition reassignment, it's possible for the Kafka connector for MicroProfile Reactive Messaging to not update its internal state correctly, resulting in it no longer receiving messages and a `NullPointerException` being logged.
```

+
The following log messages may be seen:
+
[source]
----
java.lang.NullPointerException
	at com.ibm.ws.microprofile.reactive.messaging.kafka.KafkaInput.lambda$wrapInMessageStream$10(KafkaInput.java:274)
	at com.ibm.ws.microprofile.reactive.messaging.kafka.KafkaInput$$Lambda$476/0x00000000c2be7230.test(Unknown Source)
	at io.smallrye.reactive.streams.stages.FilterStageFactory$$Lambda$481/0x00000000c3268d30.test(Unknown Source)
----
+
[source]
----
User provided listener com.ibm.ws.microprofile.reactive.messaging.kafka.adapter.impl.ConsumerRebalanceListenerImpl failed on invocation of onPartitionsRevoked for partitions [live-partition-test-in-0]
java.lang.NullPointerException
	at com.ibm.ws.microprofile.reactive.messaging.kafka.KafkaInput.onPartitionsRevoked(KafkaInput.java:346)
	at com.ibm.ws.microprofile.reactive.messaging.kafka.adapter.impl.ConsumerRebalanceListenerImpl.onPartitionsRevoked(ConsumerRebalanceListenerImpl.java:55)
	at org.apache.kafka.clients.consumer.internals.ConsumerCoordinator.invokePartitionsRevoked(ConsumerCoordinator.java:315)
----
+

> This issue has been resolved and the partition assignment succeeds correctly and without the above errors.

Reviewer: ```suggestion
This issue has been resolved and the partition assignment succeeds correctly and without these errors.
```

* link:https://github.com/OpenLiberty/open-liberty/issues/21808[Provide a way for Custom User Registries to use the uniqueId instead of the securityName]
+
When using a `CustomUserRegistry` and making a programmatic API call like `getRemoteUser`, `getCallerPrincipal` or `getUserPrincipal`, Open Liberty was incorrectly returning the `securityName`.
+

> The issue now resolved and the methods `getRemoteUser`, `getCallerPrincipal` or `getUserPrincipal` now correctly return the `displayName`.

Reviewer: ```suggestion
The issue is now resolved and the `getRemoteUser`, `getCallerPrincipal` or `getUserPrincipal` methods correctly return the `displayName`.
```

* link:https://github.com/OpenLiberty/open-liberty/issues/23252[AmbiguousResolutionException when same class is present twice and certain features are used]
+
When a class is present twice in an application and is used as a CDI managed bean and certain features are used, CDI may sometimes report an `AmbiguousResolutionException`.
+
[source]
----
Caused by: org.jboss.weld.exceptions.AmbiguousResolutionException: WELD-001318: Cannot resolve an ambiguous dependency between: 
  - Managed Bean [class com.ibm.wssvt.acme.annuity.common.business.jaxrs.MyJAXBResolver] with qualifiers [@Any @Default],
  - Managed Bean [class com.ibm.wssvt.acme.annuity.common.business.jaxrs.MyJAXBResolver] with qualifiers [@Any @Default]
----
+

> This issue is now resolved.  The same bean is no longer included twice and, therefore, the exception is no longer thrown.

Reviewer: ```suggestion
This issue is now resolved.  The same bean is no longer included twice and the exception is no longer thrown.
```

* link:https://github.com/OpenLiberty/open-liberty/issues/23225[IllegalStateException in DynaCache when app server is stopping]
+
Sometimes while the Liberty server is shutting down with the `webCache-1.0` feature enabled, an FFDC is created by a DynaCache daemon thread:
+
[source]
----
Stack Dump = java.lang.IllegalStateException: ScheduledExecutorService service is unavailable
 at com.ibm.ws.cache.Scheduler.getScheduledExecutorService(Scheduler.java:93)
 at com.ibm.ws.cache.Scheduler.createNonDeferrable(Scheduler.java:108)
 at com.ibm.ws.cache.RealTimeDaemon.alarm(RealTimeDaemon.java:107)
 at com.ibm.ws.cache.RealTimeDaemon$2.run(RealTimeDaemon.java:111)
 at com.ibm.ws.cache.Scheduler$WrappedRunnable.run(Scheduler.java:130)
 ...
----
+
The issue is now resolved.  DynaCache no longer attempts to create a schedule when the Liberty server is stopping and the FFDC is not generated.

* link:https://github.com/OpenLiberty/open-liberty/issues/23186[IdentityStore validate method not getting called for BasicAuthentication request]
+

> Some Basic Authentication requests fail due to `IdentityStore` not being called correctly.  This a result of Liberty's `BasicHttpAuthenticationMechanism` using `BasicAuthenticationCredential` instead of `UsernamePasswordCredential`. That credential object is passed down to the `IdentityStore` validate method call as a parameter.

Reviewer: ```suggestion
Some Basic Authentication requests fail due to `IdentityStore` not being called correctly.  This a result of Liberty's `BasicHttpAuthenticationMechanism` using `BasicAuthenticationCredential` instead of `UsernamePasswordCredential`. That credential object is passed down to the `IdentityStore` `validate` method call as a parameter.
```

+
The `BasicAuthenticationCredential` class extends the `UsernamePasswordCredential`.
+

> For this case, `IdentityStore` needs to have an overloaded version of `validate` method to function properly.  For example:

Reviewer: ```suggestion
For this case, `IdentityStore` needs to have an overloaded version of the `validate` method to function properly.  For example:
```

+
[source]
----
public CredentialValidationResult validate(BasicAuthenticationCredential basicAuthenticationCredential) {
        return validate(new UsernamePasswordCredential(basicAuthenticationCredential.getCaller(), basicAuthenticationCredential.getPasswordAsString()));
    }
----
+

> The issue is now resolve and the authentication succeeds with the `validate` method being called and successfully validating the user.

Reviewer: ```suggestion
The issue is now resolved and the authentication succeeds with the `validate` method being called and successfully validating the user.
```

...
----
+
The issue has been resolved and the web application now starts without the `CWWKZ0002E` error caused by the NPE.

* link:https://github.com/OpenLiberty/open-liberty/issues/22909[MDB class Java heap leak on application stop]
+
There is a reference from the Liberty message-driven bean (MDB) code to the application's MDB class through a pool that does not appear to be cleaned on application stop.  See the following reference chain:
+
[source]
----
'- enterpriseBeanClass com.ibm.ejs.container.MessageEndpointFactoryImpl @ 0x8d630f50
   '- ivMessageEnpointHandlerFactory com.ibm.ws.ejbcontainer.mdb.MessageEndpointHandlerPool @ 0x8d6307c8
      '- discardStrategy com.ibm.ws.ejbcontainer.util.PoolImplThreadSafe @ 0x8d6301e0
         '- [3] java.lang.Object[10] @ 0x8677de20
            '- elementData java.util.ArrayList @ 0x817b8f10
               '- pools com.ibm.ws.ejbcontainer.util.PoolManagerImpl @ 0x817b8ee8
                  '- poolManager com.ibm.ejs.container.EJSContainer @ 0x80f024e0
----
+
This issue is now resolved and all references to applications classes are released when an application is stopped.

> * link:https://github.com/OpenLiberty/open-liberty/issues/22865[Datasource changes are not propagating to JPA during dynamic config update]

Author: ```suggestion
* link:https://github.com/OpenLiberty/open-liberty/issues/22865[Datasource changes are not propagating to JPA during dynamic config update]
+
```

> When updating the `<datasource>` properties (`databaseName`, pas`sword, `portNumber`, `serverName`, ect) in server.xml while the server is running, the changes are not reflected in JPA applications.

Reviewer: ```suggestion
When updating the `<datasource>` properties (`databaseName`, `password`, `portNumber`, `serverName`, etc.) in the `server.xml` file while the server is running, the changes are not reflected in JPA applications.
```
